### PR TITLE
Add document for how Dask works with GPUs

### PR DESCRIPTION
**Pull request by mrocklin**
This could use review both from people who know about GPUs (@pentschev @quasiben) and from people who are unassociated with NVIDIA (to keep things honest) (@jcrist @martindurant)

I'm getting more questions about how Dask works with GPUs these days, and I'd like to have a place where I can start to consolidate information.  A page like this seems ideal to me, but I'm happy to move this elsewhere if desired.  I'm slightly concerned about the fast-moving nature of the work here.  I intentionally didn't point to more experimental projects like dask-cuda or dask-cuml or UCX.

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

### CODE REVIEW

File: docs/source/gpu.rst [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            allow="autoplay; encrypted-media"
+            allowfullscreen>
+    </iframe>
+
+
+
+High Level Collections
+----------------------
+
+Dask can also help to scale out large scale array and data computations by
```
Comment (quasiben) [inline docs/source/gpu.rst]: Small style comment on the double usage of scale

File: docs/source/gpu.rst [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+Arrays
+~~~~~~
+
+.. note:: Dask's integration with CuPy relies on features recently added to
+   NumPy and CuPy, particularly in version ``numpy>=1.17`` and ``cupy>=6``
+
+`Chainer's CuPy <https://cupy.chainer.org/>`_ library provides a GPU
+accelerated NumPy-like library that interoperates nicely with Dask Array.
+
+If you have a cupy installed then you should be able to convert a NumPy-backed
```
Comment (quasiben) [inline docs/source/gpu.rst]: remove `a`

File: docs/source/gpu.rst [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+There are a variety of GPU accelerated machine learning libraries that follow
+the Scikit-Learn Estimator API of fit, transform, and predict.  These can
+generally be used within `Dask-ML's <https://ml.dask.org>`_ meta estimators,
+such as `hyper parameter optimization <https://ml.dask.org/hyper-parameter-search.html>`_.
+
+Some of these include:
+
+-  `Skorch <https://skorch.readthedocs.io/>`_
+-  `cuML <https://rapidsai.github.io/projects/cuml/en/latest/>`_
+-  TODO (help?)
```
Comment (quasiben) [inline docs/source/gpu.rst]: You could also include 
- https://github.com/Xtra-Computing/thundergbm
- https://github.com/Xtra-Computing/thundersvm
- https://github.com/Microsoft/LightGBM
- xgboost ?

File: docs/source/gpu.rst [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+Restricting Work
+~~~~~~~~~~~~~~~~
+
+By default Dask allows as many tasks as you have CPU cores to run concurrently.
+However if your tasks primarily use a GPU then you probably want far fewer
+tasks running at once.  There are a few ways to limit parallelism here:
+
+-  Limit the number of threads explicitly on your workers using the
+   ``--nthreads`` keyword in the CLI or the ``ncores=`` keyword in Python.
```
Comment (quasiben) [inline docs/source/gpu.rst]: People may be confused by 

> keyword in `Python`.  

Suggested change to 

> ``ncores=`` keyword in the Cluster constructor

File: docs/source/gpu.rst [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+   tasks as GPU tasks so that the scheduler and limit them, while leaving the
+   rest of your CPU cores for other work
+
+Specifying GPUs per Machine
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some configurations may have many GPU devices per node.  Dask is often used to
+balance and coordinate work between these devices.
+
+In these situations it is common to start one Dask worker per device, an use
```
Comment (quasiben) [inline docs/source/gpu.rst]: `an use` -> `and use`

File: docs/source/gpu.rst [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+prefer one device.
+
+.. code-block:: bash
+
+   # If we have four GPUs on one machine
+   CUDA_VISIBLE_DEVICES=0 dask-worker ...
+   CUDA_VISIBLE_DEVICES=1 dask-worker ...
+   CUDA_VISIBLE_DEVICES=2 dask-worker ...
+   CUDA_VISIBLE_DEVICES=3 dask-worker ...
+
```
Comment (quasiben) [inline docs/source/gpu.rst]: Including a note on `dask-cuda` may also be useful.  

> Similar to Dask's `LocalCluster`, `Dask-CUDA <https://github.com/rapidsai/dask-cuda>`_  is often used for easy construction of a Dask cluster on nodes with GPUs.

**Comment by mrocklin** [comment]
Thanks for the great review @quasiben .  I think I've incorporated all of your suggestions in the most recent commit.

**Comment by quasiben** [comment]
👍

File: docs/source/gpu.rst [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+Dask DataFrame to a cuDF-backed Dask DataFrame as follows:
+
+.. code-block:: python
+
+   import cudf
+
+   df = df.map_partitions(cudf.from_pandas)  # convert pandas partitions into cudf partitions
+
+However, cuDF does not support the entire Pandas interface, and so a variety of
+Dask DataFrame operations will not function properly.
```
Comment (pentschev) [inline docs/source/gpu.rst]:
```suggestion
Dask DataFrame operations will not function properly. Check the 
`cudf API Reference <https://rapidsai.github.io/projects/cudf/en/latest/api.html>`_
for currently supported interface.
```

File: docs/source/gpu.rst [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+
+Arrays
+~~~~~~
+
+.. note:: Dask's integration with CuPy relies on features recently added to
+   NumPy and CuPy, particularly in version ``numpy>=1.17`` and ``cupy>=6``
```
Comment (pentschev) [inline docs/source/gpu.rst]: Using ``numpy>=1.16`` works too, although it's necessary to set the `NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=1` environment variable. I see that perhaps you're trying to avoid referencing the environment variable, but there are two issues to note:

1. NumPy 1.17 has not been released yet;
1. Despite the environment variable having been removed, I remember seeing some discussions that this may still be added back even for 1.17.

File: docs/source/gpu.rst [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+.. code-block:: python
+
+   import cupy
+
+   x = x.map_blocks(cupy.asarray)
+
+CuPy is fairly mature and adheres closely to the NumPy API.  However, small
+differences do exist and these can cause Dask Array operations to function
+improperly.
```
Comment (pentschev) [inline docs/source/gpu.rst]:
```suggestion
CuPy is fairly mature and adheres closely to the NumPy API.  However, small
differences do exist and these can cause Dask Array operations to function
improperly. Check the
`CuPy Reference Manual <https://docs-cupy.chainer.org/en/stable/reference/index.html>`_
for API compatibility.
```

File: docs/source/gpu.rst [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+By default Dask allows as many tasks as you have CPU cores to run concurrently.
+However if your tasks primarily use a GPU then you probably want far fewer
+tasks running at once.  There are a few ways to limit parallelism here:
+
+-   Limit the number of threads explicitly on your workers using the
+    ``--nthreads`` keyword in the CLI or the ``ncores=`` keyword the
+    Cluster constructor.
+-   Use `worker resources <https://distributed.dask.org/en/latest/resources.html>`_ and tag certain
+    tasks as GPU tasks so that the scheduler and limit them, while leaving the
```
Comment (pentschev) [inline docs/source/gpu.rst]: "... so that the scheduler and limit them, ..." -> it seems like there's something missing after "scheduler", but I'm not sure what.

Comment (pentschev) [reply docs/source/gpu.rst]: Or perhaps was it supposed to be "scheduler _will_ limit them"?

File: docs/source/gpu.rst [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+    rest of your CPU cores for other work
+
+Specifying GPUs per Machine
```
Comment (jcrist) [inline docs/source/gpu.rst]: I haven't exposed this in `dask-yarn` yet, but skein supports Hadoop 3 GPU resource requests (https://jcrist.github.io/skein/specification.html#resources), which set these environment variables automatically. If I was to expose this in `dask-yarn`, perhaps adding a link to that section from here would be appropriate?

File: docs/source/gpu.rst (hunk 8 above)
Comment (mrocklin) [reply docs/source/gpu.rst]: >  I see that perhaps you're trying to avoid referencing the environment variable

Indeed, especially for a high level document like this.

> NumPy 1.17 has not been released yet;

I'm ok with that.

> Despite the environment variable having been removed, I remember seeing some discussions that this may still be added back even for 1.17.
 
That's unfortunate.

File: docs/source/gpu.rst (hunk 10 above)
Comment (mrocklin) [reply docs/source/gpu.rst]: Indeed.  Thanks for catching this.

File: docs/source/gpu.rst (hunk 11 above)
Comment (mrocklin) [reply docs/source/gpu.rst]: Yes, that sounds good.  We should do something similar with dask-kubernetes

**Comment by pentschev** [comment]
Overall, LGTM.

**Comment by mrocklin** [comment]
Thanks all.  Merging.
